pkg/proxy/ipvs/proxier.go

```diff
@@ -720,61 +651,88 @@ func (handle *LinuxKernelHandler) GetKernelVersion() (string, error) {
 	}
-	mods = utilipvs.GetRequiredIPVSModules(kernelVersion)
-	wantModules := sets.NewString()
-	// We check for the existence of the scheduler mod and will trigger a missingMods error if not found
+
 	if scheduler == "" {
 		scheduler = defaultScheduler
 	}
-	schedulerMod := "ip_vs_" + scheduler
-	mods = append(mods, schedulerMod)
-	wantModules.Insert(mods...)
 
-	modules := wantModules.Difference(loadModules).UnsortedList()
-	var missingMods []string
-	ConntrackiMissingCounter := 0
-	for _, mod := range modules {
-		if strings.Contains(mod, "nf_conntrack") {
-			ConntrackiMissingCounter++
-		} else {
-			missingMods = append(missingMods, mod)
-		}
-	}
-	if ConntrackiMissingCounter == 2 {
-		missingMods = append(missingMods, "nf_conntrack_ipv4(or nf_conntrack for Linux kernel 4.19 and later)")
+	// Check is any virtual servers (vs) are configured. If any, we assume
+	// that this is a kube-proxy re-start and skip the checks.
+	vservers, err := ipvs.GetVirtualServers()
+	if err != nil {
+		klog.ErrorS(err, "Can't read the ipvs")
+		return err
 	}
-
-	if len(missingMods) != 0 {
-		return fmt.Errorf("IPVS proxier will not be used because the following required kernel modules are not loaded: %v", missingMods)
+	klog.V(5).InfoS("Virtual Servers", "count", len(vservers))
+	if len(vservers) > 0 {
+		klog.V(5).InfoS("Virtual server exists", "count", len(vservers))
+		// This is most likely a kube-proxy re-start. We know that ipvs works
+		// and if any VS uses the configured scheduler, we are done.
+		for _, vs := range vservers {
+			if vs.Scheduler == scheduler {
+				return nil
+			}
+		}
+		klog.V(5).InfoS("No existing VS uses the configured scheduler", "scheduler", scheduler)
+	}
+
+	// BUG: If ipvs is not compiled into the kernel ipvs.GetVirtualServers
+	// does not return an error. Instead also ipvs.AddVirtualServer returns OK
+	// (no error)!
+
+	// Try to insert a dummy VS with the passed scheduler.
+	// We should use a VIP address that is not used on the node.
+	// An address "198.51.100.0" from the TEST-NET-2 rage in https://datatracker.ietf.org/doc/html/rfc5737
+	// is used. These addresses are reserved for documentation. If the user is using
+	// this address for a VS anyway we *will* mess up, but that would be an invalid configuration.
+	// If the user have configured the address to an interface on the node (but not a VS)
+	// then traffic will temporary be routed to ipvs during the probe and dropped.
+	// The later case is also and invalid configuration, but the traffic impact will be minor.
+	// This should not be a problem if users honors reserved addresses, but cut/paste
+	// from documentation is not unheard of, so the restriction to not use the TEST-NET-2 range
+	// must be documented.
+	vs := utilipvs.VirtualServer{
+		Address:   netutils.ParseIPSloppy("198.51.100.0"),
+		Protocol:  "TCP",
+		Port:      20000,
+		Scheduler: scheduler,
+	}
+	if err := ipvs.AddVirtualServer(&vs); err != nil {
```

Comment:
What happens if we do this, then crash, then restart?  When we get back to this function, we'll find this VS exists (the above cases) and it may or may not have the correct scheduler.  If it does, we will never delete it (again in the `len(vservers) > 0` path above).

Perhaps a little extra paranoia about cleaning this magic entry up?

Reply:
There is no difference from today, the CanUseIPVSProxier() returns OK. But the kube-proxy seems to handle it. If I change the sceduler the kube-proxy re-starts. Whatever sheduler is finds if re-configures if CanUseIPVSProxier() retrurns ok. That is not part of this PR that just alters the method of determining if ipvs and the scheduler can be used, but it seem to work.

Before;
```
# ipvsadm -Ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  12.0.0.1:443 rr
  -> 192.168.1.1:6443             Masq    1      0          0         
```
then the scheduler is updated to "wlc", the kube-config detects the change and re-starts.
After;
```
# ipvsadm -Ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  12.0.0.1:443 wlc
  -> 192.168.1.1:6443             Masq    1      0          0         
```
Frankly, I don't know how it does, probably a complete re-sync, but it seem to be very solid.

Reply:
No old VS'es seem ever to linger. I can only test basic cases, but the return from CanUseIPVSProxier() is not altered, only the method of getting there.

Reply:
BTW, I tested to skip the removal of the probe-VS just to see it with "ipvsadm", but it was cleared away. I suspect that a "Flush()" is called somewhere, but I havn't investigated.

Reply:
I think we want to know why.  If I put a panic() between L730 and L731 (just before the delete), I would expect that it would restart and take the above happy-path, retruning at L686.

If not, we should document why, don't you think?

Reply:
Merry Christmas. My boy has a new desktop gaming PC and is unrechable from planet earth, so I got time for a test. I describe it in the main conversation area.

Reply:
Thanks for the explanation in the main conversation.